File: microservices/DriverService/src/test/java/com/modsen/taxi/driversrvice/AccessTokenProvider.java
package com.modsen.taxi.driversrvice;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.test.web.reactive.server.WebTestClient;

import java.util.Map;

@Service
public class AccessTokenProvider {

    private final WebTestClient webTestClient;
    private final String authServerUrl;

    @Value("${keycloak.client-id}")
    private String clientId;

    @Value("${keycloak.client-secret}")
    private String clientSecret;

    public AccessTokenProvider(WebTestClient webTestClient, @Value("${keycloak.auth-server-url}") String authServerUrl) {
        this.webTestClient = webTestClient;
        this.authServerUrl = authServerUrl;
    }

    public String getAccessToken(String username, String password) {
        String requestBody = "username=" + username + "&" +
                "password=" + password + "&" +
                "grant_type=password&" +
                "scope=openid&" +
                "client_id=" + clientId + "&" +
                "client_secret=" + clientSecret;

        Map<String, Object> responseBody = webTestClient.post()
                .uri(authServerUrl + "/realms/taxiapp-realm/protocol/openid-connect/token")
                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                .bodyValue(requestBody)
                .exchange()
                .expectStatus().isOk()
                .expectBody(Map.class)
                .returnResult()
                .getResponseBody();

        if (responseBody != null) {
            return (String) responseBody.get("access_token");
        } else {
            throw new RuntimeException("Failed to get access token");
        }
    }
}

File: microservices/PassengerService/src/test/java/com/modsen/taxi/passengerservice/AccessTokenProvider.kt
package com.modsen.taxi.passengerservice

import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.test.web.reactive.server.WebTestClient

@Service
class AccessTokenProvider(
    private val webTestClient: WebTestClient,
    @Value("\${keycloak.auth-server-url}") private val authServerUrl: String,
) {

    @Value("\${keycloak.client-id}")
    private lateinit var clientId: String

    @Value("\${keycloak.client-secret}")
    private lateinit var clientSecret: String

    fun getAccessToken(username: String, password: String): String {
        val requestBody = mapOf(
            "username" to username,
            "password" to password,
            "grant_type" to "password",
            "scope" to "openid",
            "client_id" to clientId,
            "client_secret" to clientSecret
        ).map { "${it.key}=${it.value}" }
            .joinToString("&")

        val responseBody = webTestClient.post()
            .uri("$authServerUrl/realms/taxiapp-realm/protocol/openid-connect/token")
            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
            .bodyValue(requestBody)
            .exchange()
            .expectStatus().isOk
            .expectBody(Map::class.java)
            .returnResult()
            .responseBody

        return (responseBody?.get("access_token") as? String)
            ?: throw RuntimeException("Failed to get access token")
    }
}
